python/ola/PidStoreTest.py

```diff
@@ -198,6 +198,87 @@ def testCmp(self):
     self.assertNotEqual(hash(p1b), hash(p2))
     self.assertNotEqual(hash(p1a), hash(p3))
 
+  def testPackUnpack(self):
+    store = PidStore.PidStore()
+    store.Load([os.path.join(path, "test_pids.proto")])
+
+    pid = store.GetName("DMX_PERSONALITY_DESCRIPTION")
+
+    # Pid.Pack only packs requests and Pid.Unpack only unpacks responses
+    # so test in two halves
+    args = ["42"]
+    blob = pid.Pack(args, PidStore.RDM_GET)
+    decoded = pid._requests.get(PidStore.RDM_GET).Unpack(blob)[0]
+    self.assertEqual(decoded['personality'], 42)
+
+    args = ["42", "7", "UnpackTest"]
+    blob = pid._responses.get(PidStore.RDM_GET).Pack(args)[0]
+    decoded = pid.Unpack(blob, PidStore.RDM_GET)
+    self.assertEqual(decoded['personality'], 42)
+    self.assertEqual(decoded['slots_required'], 7)
+    self.assertEqual(decoded['name'], "UnpackTest")
+
+  def testPackRanges(self):
+    store = PidStore.PidStore()
+    store.Load([os.path.join(path, "test_pids.proto")])
+
+    pid = store.GetName("REAL_TIME_CLOCK")
+
+    args = ["2020", "6", "20", "20", "20", "20"]
+    blob = pid.Pack(args, PidStore.RDM_SET)
+    self.assertTrue(len(blob) > 1)
+
+    # invalid year (2002 < 2003)
+    with self.assertRaises(PidStore.ArgsValidationError):
+      args = ["2002", "6", "20", "20", "20", "20"]
+      blob = pid.Pack(args, PidStore.RDM_SET)
+
+    # invalid month < 1
+    with self.assertRaises(PidStore.ArgsValidationError):
+      args = ["2020", "0", "20", "20", "20", "20"]
+      blob = pid.Pack(args, PidStore.RDM_SET)
+
+    # invalid month > 12
+    with self.assertRaises(PidStore.ArgsValidationError):
+      args = ["2020", "13", "20", "20", "20", "20"]
+      blob = pid.Pack(args, PidStore.RDM_SET)
+
+    # invalid month > 255
+    with self.assertRaises(PidStore.ArgsValidationError):
+      args = ["2020", "256", "20", "20", "20", "20"]
+      blob = pid.Pack(args, PidStore.RDM_SET)
+
+    # invalid negative month
+    with self.assertRaises(PidStore.ArgsValidationError):
+      args = ["2020", "-1", "20", "20", "20", "20"]
+      blob = pid.Pack(args, PidStore.RDM_SET)
+
+    # tests for string with min=max=2
+    pid = store.GetName("LANGUAGE_CAPABILITIES")
+    args = ["en"]
+    blob = pid._responses.get(PidStore.RDM_GET).Pack(args)[0]
+    self.assertTrue(len(blob) > 1)
+
+    with self.assertRaises(PidStore.ArgsValidationError):
+      args = ["e"]
+      blob = pid._responses.get(PidStore.RDM_GET).Pack(args)[0]
+
+    with self.assertRaises(PidStore.ArgsValidationError):
+      args = ["enx"]
+      blob = pid._responses.get(PidStore.RDM_GET).Pack(args)[0]
+
+    # valid empty string
+    pid = store.GetName("STATUS_ID_DESCRIPTION")
+    args = [""]
+    blob = pid._responses.get(PidStore.RDM_GET).Pack(args)[0]
+    decoded = pid.Unpack(blob, PidStore.RDM_GET)
```

Comment:
Perhaps a length check of the blob here too? Check it's not adding null termination or anything.

Reply:
added the check to make sure it's not there, though the python code, at least, rstrips nulls from strings

Reply:
Thanks, yeah that was just a random example.